internal/data/metric.go

```diff
@@ -0,0 +1,629 @@
+	exemplar HistogramBucketExemplar
+}
+
+func NewHistogramBucket() *HistogramBucket {
+	return &HistogramBucket{orig: &otlpmetric.HistogramValue_Bucket{}}
+}
+
+// NewHistogramBucketSlice creates a slice of HistogramBucket that are correctly initialized.
+func NewHistogramBucketSlice(len int) []HistogramBucket {
+	origs := make([]otlpmetric.HistogramValue_Bucket, len)
+	wrappers := make([]HistogramBucket, len)
+	for i := range origs {
+		wrappers[i].orig = &origs[i]
+	}
+	return wrappers
+}
+
+func (hb *HistogramBucket) Count() uint64 {
+	// TODO: Remove conversion when proto and the generated code is fixed.
+	return uint64(hb.orig.Count)
+}
+
+func (hb *HistogramBucket) SetCount(v uint64) {
+	// TODO: Remove conversion when proto and the generated code is fixed.
+	hb.orig.Count = int64(v)
+}
+
+func (hb *HistogramBucket) Exemplar() HistogramBucketExemplar {
+	return hb.exemplar
+}
+
+func (hb *HistogramBucket) SetExemplar(v HistogramBucketExemplar) {
+	hb.exemplar = v
+}
+
+// HistogramBucket are example points that may be used to annotate aggregated Histogram values.
+// They are metadata that gives information about a particular value added to a Histogram bucket.
+//
+// Must use NewHistogramBucketExemplar* functions to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type HistogramBucketExemplar struct {
+	// Wrap OTLP HistogramValue_Bucket_Exemplar.
+	orig *otlpmetric.HistogramValue_Bucket_Exemplar
+
+	// Override a few fields. These fields are the source of truth. Their counterparts
+	// stored in corresponding fields of "orig" are ignored.
+	attachments Labels
+}
+
+func NewHistogramBucketExemplar() *HistogramBucketExemplar {
+	return &HistogramBucketExemplar{orig: &otlpmetric.HistogramValue_Bucket_Exemplar{}}
+}
+
+func (hbe *HistogramBucketExemplar) Value() float64 {
+	return hbe.orig.Value
+}
+
+func (hbe *HistogramBucketExemplar) SetValue(v float64) {
+	hbe.orig.Value = v
+}
+
+func (hbe *HistogramBucketExemplar) Timestamp() TimestampUnixNano {
+	return TimestampUnixNano(hbe.orig.TimestampUnixnano)
+}
+
+func (hbe *HistogramBucketExemplar) SetTimestamp(v TimestampUnixNano) {
+	hbe.orig.TimestampUnixnano = uint64(v)
+}
+
+func (hbe *HistogramBucketExemplar) Attachments() Labels {
+	return hbe.attachments
+}
+
+func (hbe *HistogramBucketExemplar) SetAttachments(v Labels) {
+	hbe.attachments = v
+}
+
+// SummaryDataPoint is a single data point in a timeseries that describes the time-varying
+// values of a Summary metric.
+//
+// Must use NewSummaryDataPoint* functions to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type SummaryDataPoint struct {
+	// Wrap OTLP SummaryDataPoint.
+	orig *otlpmetric.SummaryDataPoint
+
+	// Override a few fields. These fields are the source of truth. Their counterparts
+	// stored in corresponding fields of "orig" are ignored.
+	labels             Labels
+	valueAtPercentiles []*SummaryValueAtPercentile
```

Comment:
I suggest we store SummaryValueAtPercentile by value, i.e.
```suggestion
	valueAtPercentiles []SummaryValueAtPercentile
```
This is similar to AttributeValue and will be more efficient.

Reply:
I would rather to a new PR that checks all this possibilities, arguably all entries should use values instead of pointers. Are you ok with that?

Reply:
>checks all these possibilities

Do you mean to benckmark or something else where you say "check"?

Reply:
I feel that we were not consistent in determining where we have values vs pointers in the new data model.

Reply:
It is often dictated by performance considerations. Alternatively you pass everything by value and then use pimpl pattern internally if passing by value turns out to be too slow.